### PR TITLE
Fixing bug where initial subscription failure causes shard consumer to get stuck.

### PR DESCRIPTION
**Pull request by yatins47**
Fixing bug where initial subscription failure causes shard consumer to get stuck.

*Issue #, if available:*

*Description of changes:*
This change is to fix the scenario where the initial SubscribeToShard call failing silently in the SDK layer during the state transition and KCL will be not be able to set the variables which causes the health check to retry the failure. The shard consumer will hold the lease and get stuck without making any progress. 
The fix is to update the lastRequestTime variable when we start the subscription to get the health check work in case the initial SubscribeToShard call fail and we don't make progress. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java [hunk 1]
```diff
@@ -448,6 +520,31 @@ public void cancel() {
...
+        private int subscriptionTryCount = 0;
+
+        @Override
+        public void subscribe(Subscriber<? super RecordsRetrieved> s) {
+            subscriber = s;
+            ++subscriptionTryCount;
+            s.onSubscribe(new Subscription() {
+                @Override
+                public void request(long n) {
+                    if (subscriptionTryCount == 1) {
```
Comment (micah-jaffe) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: Would it be simpler to rewrite this block as just `if (subscriptionTryCount != 1)`?

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java [hunk 2]
```diff
@@ -311,6 +309,80 @@ public void restartAfterRequestTimerExpiresTest() throws Exception {
...
+        synchronized (processedNotifier) {
+            subscriber.startSubscriptions();
+            processedNotifier.wait(1000);
+        }
+
+        Stream.iterate(2, i -> i + 1).limit(97).forEach(this::addUniqueItem);
+
+        addTerminalMarker(2);
+
+        verify(shardConsumer, times(0)).handleInput(argThat(eqProcessRecordsInput(processRecordsInput)),
```
Comment (micah-jaffe) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: If you're trying to verify that the shardConsumer doesn't initialize successfully, might it be clearer to use the `never()` matcher rather than `timer(0)`?

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java [hunk 3]
```diff
@@ -311,6 +309,80 @@ public void restartAfterRequestTimerExpiresTest() throws Exception {
...
+            processedNotifier.wait(5000);
+        }
+
+        barrier.await(500, TimeUnit.MILLISECONDS);
+
+        verify(shardConsumer, times(100)).handleInput(argThat(eqProcessRecordsInput(processRecordsInput)),
+                any(Subscription.class));
+
+        assertThat(received.size(), equalTo(recordsPublisher.responses.size()));
+        Stream.iterate(0, i -> i + 1).limit(received.size()).forEach(i -> assertThat(received.get(i),
```
Comment (micah-jaffe) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: Could you add a comment for what this test is verifying? It's a little hard to follow

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java [hunk 4]
```diff
@@ -311,6 +309,80 @@ public void restartAfterRequestTimerExpiresTest() throws Exception {
 
     }
 
+    @Test
+    public void restartAfterRequestTimerExpiresAfterInitialSubscriptionFailsTest() throws Exception {
```
Comment (micah-jaffe) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: How long does this test take to run? Generally we want to avoid timing-related tests as it increases the build time, but in this case that might be the best way to go about it (not close enough to the issue to say)

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java [hunk 5]
```diff
@@ -73,6 +73,7 @@
 
     void startSubscriptions() {
         synchronized (lockObject) {
+            lastRequestTime = Instant.now();
```
Comment (ashwing) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java]: 1. Can you add a comment on why we setting this here? 
2. Also add a variable comment on lastRequestTime (Last time an upstream request(N) call made to the service)
3. Refactor the places where lastRequestTime = Instant.now() followed by startSubscriptions()
4. Check why shouldn't we set this variable on call request(N) calls made from this class.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java (hunk 4 above)
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: The test case is hard to follow and has lot of waits. We need to avoid sleeps. adding comments would be helpful. We can sync up offline to check if the test can be shortened.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java (hunk 1 above)
Comment (yatins47) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: Changed to what u suggested. Agree it seems simpler to read.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java (hunk 2 above)
Comment (yatins47) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: i didn't knew about never() , new to mockito always been a EasyMock user :)

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java (hunk 3 above)
Comment (yatins47) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: Added comments and simplifying the test too.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java (hunk 4 above)
Comment (yatins47) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: I simplified the test. It takes around 60 to 80 ms to run in general which is in line with other tests in the class.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java (hunk 5 above)
Comment (yatins47) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java]: Added comments. Other points as discussed offline.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java (hunk 4 above)
Comment (yatins47) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: I have refactored the unit test and added comments to made it easier to read. It takes for it to 60 to 80 ms to run.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java [hunk 6]
```diff
@@ -311,6 +310,64 @@ public void restartAfterRequestTimerExpiresTest() throws Exception {
...
+
+        addTerminalMarker(2);
+
+        // Doing the health check to allow the subscription to restart.
+        assertThat(subscriber.healthCheck(1), nullValue());
+
+        // Allow time for processing of the records to end in the executor thread which call notifyAll as it gets the
+        // terminal record.
+        synchronized (processedNotifier) {
+            processedNotifier.wait(500);
```
Comment (micah-jaffe) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: Is 500 millis guaranteed to be enough time to reach the terminal record, even under conditions of high CPU usage or other potential slowdowns? We don't want a test failure based on internals we have no control over. 

Would mocking your executorService and telling it explicitly what to do on each call let you avoid this wait?

Comment (yatins47) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: The test here is actually about sending records from RecordPublishers to the Subscriber through the executor service. The avg time taken to receive all the records from the record publisher is around 60-80 ms hence 500 seems to be a good general timeout. I can update it to a higher value since this is a timeout and having it higher should not affect our test run time.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java [hunk 7]
```diff
@@ -44,6 +44,7 @@
 
     @VisibleForTesting
     final Object lockObject = new Object();
+    // This holds the last time a request to upstream service is made including the first try to establish subscription.
```
Comment (ashwing) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java]: actually it is an attempt to request

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java (hunk 5 above)
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java]: Line 139 needs to be removed.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java [hunk 8]
```diff
@@ -311,6 +310,64 @@ public void restartAfterRequestTimerExpiresTest() throws Exception {
...
 
+    @Test
+    public void restartAfterRequestTimerExpiresAfterInitialSubscriptionFailsTest() throws Exception {
+
+        executorService = Executors.newFixedThreadPool(1, new ThreadFactoryBuilder()
+                .setNameFormat("test-" + testName.getMethodName() + "-%04d").setDaemon(true).build());
+
+        // Mock record publisher which doesn't publish any records on first try which simulates any scenario which
+        // causes first subscription try to fail.
+        recordsPublisher = new RecordPublisherWithInitialFailureSubscription();
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: Test case looks fine. But can we simulate the failure by mocking the executor service rather than the FailingPublisher, as that closely resembles the failure that we are actually trying to address. It should be a simple change though.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java (hunk 7 above)
Comment (yatins47) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriber.java]: Modified the wording

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java (hunk 8 above)
Comment (yatins47) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java]: As discussed offline, added a new test case specifically for Rejection execution exception from executor service.
